bindings/bind_py_structure_manager.cc

```diff
@@ -458,6 +458,132 @@ namespace rascal {
+            keys_list.append(t_key);
+            // feature_dict[t_key] = features;
+          }
+
+          int i_key{0};
+          for (auto & key : all_keys) {
+            int i_center_global{0};
+            auto t_key{keys_list[i_key]};
+            // auto & feat_global = feature_dict[t_key];
+            for (auto & manager : managers) {
+              auto && property =
+                  manager->template get_property_ref<Prop_t>(property_name);
+              // size_t n_center{property.size()};
+
+              for (auto center : manager) {
+                auto && prop_row = property[center];
+                if (prop_row.count(key) == 1) {
+                  // get the feature and flatten the array
+                  auto feat_row = VecMap_t(prop_row[key].data(), inner_size);
+                  features.row(i_center_global) = feat_row;
+                }
+                i_center_global++;
+              }
+            }
+            feature_dict[t_key] = features;
+            features.setZero();
+            ++i_key;
+          }
+          return feature_dict;
+        });
+    manager_collection.def(
+        "get_dense_feature_matrix",
+        [](ManagerCollection_t & managers, const Calculator & calculator,
+           py::list & all_keys_l) {
+          using Manager_t = typename ManagerCollection_t::Manager_t;
+          using Prop_t = typename Calculator::template Property_t<Manager_t>;
+          using Keys_t = typename Prop_t::Keys_t;
+          Keys_t all_keys;
+          // convert the list of keys from python in to the proper type
+          for (py::handle key_l : all_keys_l) {
+            auto key = py::cast<std::vector<int>>(key_l);
+            all_keys.insert(key);
+          }
+
+          auto property_name{managers.get_calculator_name(calculator, false)};
+
+          auto && property_ =
+              managers[0]->template get_property_ref<Prop_t>(property_name);
+          // assume inner_size is consistent for all managers
+          int inner_size{property_.get_nb_comp()};
+
+          math::Matrix_t features{};
+
+          auto n_rows{managers.get_number_of_elements(calculator, false)};
+          size_t n_cols{all_keys.size() * inner_size};
+          features.resize(n_rows, n_cols);
+          features.setZero();
+
+          int i_row{0};
+          for (auto & manager : managers) {
+            auto && property =
+                manager->template get_property_ref<Prop_t>(property_name);
+            auto n_rows_manager = property.size();
+            property.fill_dense_feature_matrix(
+                features.block(i_row, 0, n_rows_manager, n_cols), all_keys);
+            i_row += n_rows_manager;
+          }
+
+          return features;
+        },
+        R"(Get the dense feature matrix associated with the calculator and
+        the collection of structures (managers) using the list of keys
+        provided. Only applicable when Calculator uses BlockSparseProperty.)");
```

Comment:
I do that in structure_manager_collection.hh with the FeatureMatrixHelper class. Feels too much for a lambda function. You will get a compilation error if you use it for CalculatorSortedCoulomb since there is an explicit use of Keys_t which does not exist in Property.

Reply:
`static_assert<std::is_same<Property, BlockSparseProperty>>`?